File: web/types/knowledge.ts
/**
 * Knowledge Base types for the frontend.
 */

// ==================== Knowledge Base Types ====================

export interface KnowledgeBase {
  id: string;
  name: string;
  description: string;
  collection_name: string;
  embedding_model: string;
  document_count: number;
  chunk_count: number;
  status: KnowledgeBaseStatus;
  error_message: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export type KnowledgeBaseStatus = "active" | "indexing" | "error";

export interface KnowledgeBaseListResponse {
  knowledge_bases: KnowledgeBase[];
  total: number;
}

export interface CreateKnowledgeBaseRequest {
  name: string;
  description?: string;
  embedding_model?: string;
}

export interface UpdateKnowledgeBaseRequest {
  name?: string;
  description?: string;
}

// ==================== Document Types ====================

export interface KnowledgeDocument {
  id: string;
  knowledge_base_id: string;
  filename: string;
  file_type: string;
  file_size: number;
  chunk_count: number;
  status: DocumentStatus;
  error_message: string | null;
  created_at: string;
  indexed_at: string | null;
  // Metadata fields
  tags: string[];
  category: string | null;
  author: string | null;
  custom_metadata: Record<string, unknown>;
}

export interface UpdateDocumentMetadataRequest {
  tags?: string[];
  category?: string;
  author?: string;
  custom_metadata?: Record<string, unknown>;
}

export interface DocumentTag {
  tag: string;
  usage_count: number;
}

export interface DocumentTagListResponse {
  tags: DocumentTag[];
  total: number;
}

export interface DocumentCategoryListResponse {
  categories: string[];
  total: number;
}

export type DocumentStatus = "pending" | "processing" | "indexed" | "error";

export interface KnowledgeDocumentListResponse {
  documents: KnowledgeDocument[];
  total: number;
}

export interface UploadDocumentResponse {
  document: KnowledgeDocument;
  message: string;
}

export interface BatchUploadResponse {
  documents: KnowledgeDocument[];
  success_count: number;
  error_count: number;
  errors: string[];
}

// ==================== Search Types ====================

export interface SearchKnowledgeBaseRequest {
  query: string;
  top_k?: number;
  similarity_threshold?: number;
}

export interface SearchResultItem {
  content: string;
  score: number;
  metadata: Record<string, unknown>;
}

export interface SearchKnowledgeBaseResponse {
  results: SearchResultItem[];
  query: string;
  total_results: number;
}

// ==================== UI Helper Types ====================

export interface KnowledgeBaseFormData {
  name: string;
  description: string;
  embedding_model: string;
}

export const DEFAULT_KB_FORM_DATA: KnowledgeBaseFormData = {
  name: "",
  description: "",
  embedding_model: "text-embedding-3-small",
};

export const SUPPORTED_FILE_TYPES = [
  "pdf", "docx", "doc", "pptx", "ppt",
  "xlsx", "xls", "csv",
  "txt", "md", "markdown", "html", "htm",
  "jpg", "jpeg", "png", "gif", "bmp", "tiff", "tif", "webp"
] as const;
export type SupportedFileType = (typeof SUPPORTED_FILE_TYPES)[number];

export const FILE_TYPE_LABELS: Record<SupportedFileType, string> = {
  pdf: "PDF Document",
  docx: "Word Document",
  doc: "Word Document (Legacy)",
  pptx: "PowerPoint",
  ppt: "PowerPoint (Legacy)",
  xlsx: "Excel Spreadsheet",
  xls: "Excel (Legacy)",
  csv: "CSV Data",
  txt: "Text File",
  md: "Markdown",
  markdown: "Markdown",
  html: "HTML Page",
  htm: "HTML Page",
  jpg: "JPEG Image",
  jpeg: "JPEG Image",
  png: "PNG Image",
  gif: "GIF Image",
  bmp: "Bitmap Image",
  tiff: "TIFF Image",
  tif: "TIFF Image",
  webp: "WebP Image",
};

// File type categories for preview handling
export type FileTypeCategory = "pdf" | "image" | "text" | "spreadsheet" | "presentation" | "document";

export function getFileTypeCategory(fileType: string): FileTypeCategory {
  const type = fileType.toLowerCase();
  if (type === "pdf") return "pdf";
  if (["jpg", "jpeg", "png", "gif", "bmp", "tiff", "tif", "webp"].includes(type)) return "image";
  if (["txt", "md", "markdown", "html", "htm"].includes(type)) return "text";
  if (["xlsx", "xls", "csv"].includes(type)) return "spreadsheet";
  if (["pptx", "ppt"].includes(type)) return "presentation";
  return "document";
}

export const MAX_FILE_SIZE = 50 * 1024 * 1024; // 50MB (increased for larger documents)

export function formatFileSize(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

export function getDocumentStatusLabel(status: DocumentStatus): string {
  switch (status) {
    case "pending":
      return "Pending";
    case "processing":
      return "Processing";
    case "indexed":
      return "Indexed";
    case "error":
      return "Error";
  }
}

export function getDocumentStatusColor(
  status: DocumentStatus
): "default" | "secondary" | "destructive" | "outline" {
  switch (status) {
    case "pending":
      return "outline";
    case "processing":
      return "secondary";
    case "indexed":
      return "default";
    case "error":
      return "destructive";
  }
}

File: src/api/schemas/knowledge.py
"""API schemas for knowledge base management."""
from datetime import datetime
from enum import Enum
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class SearchModeEnum(str, Enum):
    """Search mode for knowledge base queries."""
    VECTOR = "vector"  # Pure semantic search
    BM25 = "bm25"  # Pure lexical search (with vector pre-filter)
    HYBRID = "hybrid"  # Combined vector + BM25 with RRF
    HYBRID_RERANK = "hybrid_rerank"  # Hybrid + cross-encoder re-ranking


# ==================== Knowledge Base Schemas ====================


class CreateKnowledgeBaseRequest(BaseModel):
    """Request to create a new knowledge base."""

    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field("", max_length=1000)
    embedding_model: str = Field("text-embedding-3-small", max_length=100)


class UpdateKnowledgeBaseRequest(BaseModel):
    """Request to update a knowledge base."""

    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)


class KnowledgeBaseResponse(BaseModel):
    """Response containing knowledge base details."""

    id: str
    name: str
    description: str
    collection_name: str
    embedding_model: str
    document_count: int
    chunk_count: int
    status: str
    error_message: Optional[str]
    created_by: Optional[str]
    created_at: datetime
    updated_at: datetime


class KnowledgeBaseListResponse(BaseModel):
    """Response containing list of knowledge bases."""

    knowledge_bases: List[KnowledgeBaseResponse]
    total: int


# ==================== Document Schemas ====================


class KnowledgeDocumentResponse(BaseModel):
    """Response containing document details."""

    id: str
    knowledge_base_id: str
    filename: str
    file_type: str
    file_size: int
    chunk_count: int
    status: str
    error_message: Optional[str]
    created_at: datetime
    indexed_at: Optional[datetime]
    # Metadata fields
    tags: List[str] = []
    category: Optional[str] = None
    author: Optional[str] = None
    custom_metadata: dict = {}


class KnowledgeDocumentListResponse(BaseModel):
    """Response containing list of documents."""

    documents: List[KnowledgeDocumentResponse]
    total: int


class UploadDocumentResponse(BaseModel):
    """Response after uploading a document."""

    document: KnowledgeDocumentResponse
    message: str


class BatchUploadResponse(BaseModel):
    """Response after uploading multiple documents."""

    documents: List[KnowledgeDocumentResponse]
    success_count: int
    error_count: int
    errors: List[str]


# ==================== Metadata Schemas ====================


class UpdateDocumentMetadataRequest(BaseModel):
    """Request to update document metadata."""

    tags: Optional[List[str]] = Field(None, max_length=20)
    category: Optional[str] = Field(None, max_length=100)
    author: Optional[str] = Field(None, max_length=200)
    custom_metadata: Optional[dict] = None


class DocumentTagResponse(BaseModel):
    """A tag used in the knowledge base."""

    tag: str
    usage_count: int


class DocumentTagListResponse(BaseModel):
    """Response containing list of tags for autocomplete."""

    tags: List[DocumentTagResponse]
    total: int


class DocumentCategoryListResponse(BaseModel):
    """Response containing list of categories."""

    categories: List[str]
    total: int


# ==================== Search Schemas ====================


class SearchKnowledgeBaseRequest(BaseModel):
    """Request to search a knowledge base."""

    query: str = Field(..., min_length=1, max_length=1000)
    top_k: int = Field(5, ge=1, le=20)
    similarity_threshold: float = Field(0.1, ge=0.0, le=1.0)
    mode: SearchModeEnum = Field(
        SearchModeEnum.VECTOR,
        description="Search mode: vector (semantic), bm25 (lexical), hybrid, or hybrid_rerank"
    )
    rerank: bool = Field(
        False,
        description="Enable cross-encoder re-ranking (applies to hybrid mode)"
    )


class SearchResultItem(BaseModel):
    """A single search result."""

    content: str
    score: float
    metadata: dict


class SearchKnowledgeBaseResponse(BaseModel):
    """Response from knowledge base search."""

    results: List[SearchResultItem]
    query: str
    total_results: int
